libyul/optimiser/UnusedFunctionParameterPruner.cpp

```diff
@@ -0,0 +1,136 @@
+{
+
+bool anyFalse(vector<bool> const& _mask)
+{
+	return any_of(_mask.begin(), _mask.end(), [](bool b){ return !b; });
+}
+
+template <typename K, typename V>
+optional<V> optionalFind(map<K, V> const& _map, K const& _key)
+{
+	auto it = _map.find(_key);
+
+	if (it != _map.end())
+		return it->second;
+
+	return nullopt;
+}
+
+
+} // anonymous namespace
+
+void UnusedFunctionParameterPruner::run(OptimiserStepContext& _context, Block& _block)
+{
+
+	map<YulString, size_t> references = ReferencesCounter::countReferences(_block);
+	auto used = [&](auto v) -> bool { return references.count(v.name); };
+
+	// Function name and a boolean mask, where `false` at index `i` indicates that the function
+	// argument at index `i` in `FunctionDefinition::parameters` is unused inside the function body.
+	map<YulString, vector<bool>> usedParameters;
+	// Function name and a boolean mask, where `false` at index `i` indicates that the
+	// return-parameter at index `i` in `FunctionDefinition::returnVariables` is unused inside the
+	// function body.
+	map<YulString, vector<bool>> usedReturnVariables;
+
+	// Step 1 of UnusedFunctionParameterPruner: Find functions whose parameters (both arguments and
+	// return-parameters) are not used in its body.
+	for (auto const& statement: _block.statements)
+		if (holds_alternative<FunctionDefinition>(statement))
+		{
+			FunctionDefinition const& f = std::get<FunctionDefinition>(statement);
+
+			if (wasPruned(f, _context.dialect))
+				continue;
+
+			vector<bool> parameters = applyMap(f.parameters, used);
+			vector<bool> returnVariables = applyMap(f.returnVariables, used);
+
+			if (anyFalse(parameters))
+				usedParameters[f.name] = move(parameters);
+			if (anyFalse(returnVariables))
+				usedReturnVariables[f.name] = move(returnVariables);
+		}
+
+	set<YulString> namesToFree = util::keys(usedParameters) + util::keys(usedReturnVariables);
+
+	// Step 2 of UnusedFunctionParameterPruner: Replace all references of the function, say `f`, by
+	// a new name, say `f_1`.
+	NameDisplacer replace{_context.dispenser, namesToFree};
+	replace(_block);
+	// Inverse-Map of the above translations. In the above example, this will store an element with
+	// key `f_1` and value `f`.
+	std::map<YulString, YulString> newToOriginalNames = invertMap(replace.translations());
+
+	// Step 3 of UnusedFunctionParameterPruner: introduce a new function in the block with body of
+	// the old one. Replace the body of the old one with a function call to the new one with reduced
+	// parameters.
+	//
+  	// For example: introduce a new function `f` with the same the body as `f_1`, but with reduced
+	// parameters, i.e., `function f() -> y { y := 1 }`. Now replace the body of `f_1` with a call to
+	// `f`, i.e., `f_1(x) -> y { y := f() }`.
+	iterateReplacing(_block.statements, [&](Statement& _s) -> optional<vector<Statement>>
+	{
+		if (holds_alternative<FunctionDefinition>(_s))
+		{
+			FunctionDefinition& old = get<FunctionDefinition>(_s);
```

Comment:
Will `old` or `replacement` be the linking function?

Reply:
I changed the name of `old` to `originalFunctionWithNewName` and the function `createReplacement` to `createLinkingFunction`.